### ISSUE TITLE
LDAP integration with MTLS is failing

### ISSUE THREAD

**Issue by pravsjv** [keycloak/keycloak#26423]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

core

### Describe the bug

Hi Team,
We are trying to integrate LDAP in MTLS mode and were not able to successfully integrate it. Every time we enable startTLS button, the communication fails. When tried with the LDAP search query using the same certs integrated with keycloak, the query returns the data. Also, we implemented a Java client to query LDAP using the same certs, it's working fine.
So we tried to check open-source GitHub discussions and stories but looks like there is no success anywhere.

Errors:
654b4791 conn=13302 fd=12 ACCEPT from IP=IP:PORT (IP=0.0.0.0:PORT)
TLS: can't accept: A TLS fatal alert has been received..
654b4792 conn=13302 fd=12 closed (TLS negotiation failure)

"level":"ERR","log":{"message":"{"host":"ckey1-ckey-0","version":"1.1","short_message":"KC-SERVICES0055: Error when authenticating to LDAP: LDAP connection has been closed","full_message":"KC-SERVICES0055: Error when authenticating to LDAP: LDAP connection has been closed","timestamp":1.699466713409E9,"level":3,"facility":"keycloak","_LoggerName":"org.keycloak.services","_SourceSimpleClassName":"LDAPServerCapabilitiesManager","_StackTrace":"javax.naming.NamingException: LDAP connection has been closed\n\tat java.naming/com.sun.jndi.ldap.LdapRequest.getReplyBer(Unknown Source)\n\tat java.naming/com.sun.jndi.ldap.Connection.readReply(Unknown Source)\n\tat java.naming/com.sun.jndi.ldap.LdapClient.ldapBind(Unknown Source)\n\tat java.naming/com.sun.jndi.ldap.LdapClient.authenticate(Unknown Source)\n\tat java.naming/com.sun.jndi.ldap.LdapCtx.connect(Unknown Source)\n\tat java.naming/com.sun.jndi.ldap.LdapCtx.(Unknown Source)\n\tat java.naming/com.sun.jndi.ldap.LdapCtxFactory.getLdapCtxFromUrl(Unknown Source)\n\tat java.naming/com.sun.jndi.ldap.LdapCtxFactory.getUsingURL(Unknown Source)\n\tat java.naming/com.sun.jndi.ldap.LdapCtxFactory.getUsingURLs(Unknown Source)\n\tat java.naming/com.sun.jndi.ldap.LdapCtxFactory.getLdapCtxInstance(Unknown Source)\n\tat java.naming/com.sun.jndi.ldap.LdapCtxFactory.getInitialContext(Unknown Source)\n\tat java.naming/javax.naming.spi.NamingManager.getInitialContext(Unknown Source)\n\tat java.naming/javax.naming.InitialContext.getDefaultInitCtx(Unknown Source)\n\tat java.naming/javax.naming.InitialContext.init(Unknown Source)\n\tat java.naming/javax.naming.ldap.InitialLdapContext.(Unknown Source)\n\tat org.keycloak.storage.ldap.idm.store.ldap.LDAPContextManager.createLdapContext(LDAPContextManager.java:80)\n\tat org.keycloak.storage.ldap.idm.store.ldap.LDAPContextManager.getLdapContext(LDAPContextManager.java:100)\n\tat org.keycloak.services.managers.LDAPServerCapabilitiesManager.testLDAP(LDAPServerCapabilitiesManager.java:139)\n\tat org.keycloak.services.resources.admin.TestLdapConnectionResource.testLDAPConnection(TestLdapConnectionResource.java:93)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)\n\tat java.base/java.lang.reflect.Method.invoke(Unknown Source)\n\tat org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:154)\n\tat org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:118)\n\tat org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:560)\

### Version

23.0.4

### Expected behavior

When enabling MTLS button, LDAP integration throws an error

### Actual behavior

LDAP integration with MTLS is expected to work.

### How to Reproduce?

Integrate LDAP with Keycloak and enable MTLS button.

### Anything else?

We have been trying to integrate keycloak with LDAP in MTLS mode for some time and none of them works. Can team look into this and provide us any guideline or a document if they are successful in the integration as individual java client with MTLS is working fine but not with keycloak.

**Comment by pravsjv** [comment]
@sguilhen @pedroigor The following doc is to enable MTLS in keycloak and this configuration works fine. The problem is when we configure LDAP with TLS(Enable TLS in keycloak LDAP integration page), we observe the above issues as mentioned in the description. The same integration and configuration when tested with a standalone java client in TLS communication it works fine.

**Comment by pedroigor** [comment]
@pravsjv Using `ldaps` should work fine if you have configured the [truststore](https://www.keycloak.org/server/keycloak-truststore) properly.

I'm not sure what you mean by MTLS. Are you talking about `StartTLS` instead? If so, make sure you connect to Keycloak using the `ldap` scheme (not `ldaps`).

**Comment by HaritSharma16** [comment]
@pedroigor By MTLS we mean to say (TWO WAY SSL), which mean LDAP server is configured in such a way that it has enabled client authentication, means keycloak will verify the LDAP's server certificates(LDAP server's server certificates) using the CA certificate which we will keep in Keycloak's truststore(This CA should be the one which has signed the LDAP's server certificates) and LDAP is also expecting that keycloak should present the client certificates (containing public cert and private key) while connecting and it will verify the certificate presented by keycloak.

So in this configuration we have a server configured, which are able to connect to and also query using simple standalone java application.

But via keycloak we are not able to get positive result from **Test Authentication**  button and it fails with the above mentioned error.

**Comment by pravsjv** [comment]
@sguilhen We(@HaritSharma16 ) have provided the required info as requested, but still the JIRA is closed. We have opened the same requirement multiple times and is closed without any action for the past 4-5 releases. Could you please reopen and provide some help on this issue.

**Comment by pedroigor** [comment]
@pravsjv I think we are missing how to reproduce the problem. When closing this issue I was able to connect using `ldaps` after setting up the truststore. See https://github.com/keycloak/keycloak/issues/26423#issuecomment-2018787101.

Or Am I missing something else?

Perhaps we ca re-start the discussions by having your ldap provider configuration so we make sure we are talking about the same thing?

**Comment by pedroigor** [comment]
@HaritSharma16 Sorry for the delay.

After reading all the discussions we had, mainly your last comment and this one https://github.com/keycloak/keycloak/issues/26423#issuecomment-2018787101, I think we are indeed talking about different things.

As per the issue description, you are facing errors when using StartTLS. This protocol command is an extension to plain text communication protocols, which offers a way to upgrade a plain text connection to an encrypted (TLS or SSL) connection instead of using a separate port for encrypted communication.

This should work fine as long as you have a trust store set and are using `ldap`.

AFAIK, we do not support MTLS so that Keycloak will send its certificate to *authenticate* to the server. I would ask @mposolda or @rmartinc if they think otherwise.

**Comment by rmartinc** [comment]
@pedroigor We don't have any documentation to configure mTLS in LDAP, so I would say that it's not supported. Using `Use Truststore SPI` as `Always`, it's sure it won't work because the `SSLSocketFactory` does not have any `KeyManager` (only `TrustManager`). Maybe using `Never` it can work, it defaults to java socket factory. But as you commented this is not documented, supported or tested.

**Comment by pedroigor** [comment]
Closing as we do not support mTLS when authenticating to an LDAP server.

**Comment by pravsjv** [comment]
@pedroigor can this requirement be taken as a new feature request or new enhancement?

**Comment by sguilhen** [comment]
@pravsjv I don't see any problem in raising an enhancement request for this

**Comment by durgasusm** [comment]
We were able to get Keycloak working with mTLS-enabled LDAP by adjusting the Truststore SPI setting and by passing Java arguments.

- When we switched to Use Truststore SPI = Never, Keycloak correctly picked up the keystore settings provided via Java system properties:
`-Djavax.net.ssl.keyStore=<keystore_file> \
-Djavax.net.ssl.keyStorePassword=<password>
`
With this combination, mTLS authentication to LDAP works as expected.

- The default setting When Use Truststore SPI = Always, Keycloak only sends truststore information and does not use the client certificate from the keystore and Keycloak doesnot connect with mTLS enabled LDAP server.


Suggestion: It would be helpful if Keycloak supported configuring both truststore and keystore for LDAP directly via the SPI (instead of relying on JVM arguments). This would allow administrators to manage mTLS entirely within Keycloak configuration rather than mixing with JVM-level system properties.
